Carsharing Aggregator/Application/Auth/Login/View/LoginView.swift

```diff
@@ -0,0 +1,318 @@
+        }
+        textField.snp.remakeConstraints { make in
+            make.top.equalTo(previousTextField.snp.bottom).offset(offset)
+            make.centerX.equalTo(snp.centerX)
+            make.height.equalTo(previousTextField.snp.height)
+            make.width.equalTo(previousTextField.snp.width)
+            
+        }
+    }
+        
+    private func updateConstraintsForValidTextField(_ textField: UITextField, relativeTo previousTextField: UITextField?, isValid: Bool) {
+        let previousTextField = previousTextField ?? UITextField()
+        let offset = isValid ? 16 : 53
+        
+        if !isValid {
+            previousTextField.layer.borderColor = UIColor.red.cgColor
+        } else {
+            previousTextField.layer.borderColor = UIColor.black.cgColor
+        }
+        textField.snp.remakeConstraints { make in
+            make.top.equalTo(previousTextField.snp.bottom).offset(offset)
+            make.centerX.equalTo(snp.centerX)
+            make.height.equalTo(previousTextField.snp.height)
+            make.width.equalTo(previousTextField.snp.width)
+        }
+    }
+    
+    private func updateConstraintsForLabel(_ view: UIView, relativeTo previousTextField: UITextField, isEmpty: Bool, offset: CGFloat) {
+        if isEmpty {
+            previousTextField.layer.borderColor = UIColor.red.cgColor
+        } else {
+            previousTextField.layer.borderColor = UIColor.black.cgColor
+        }
+        view.snp.remakeConstraints { make in
+            make.top.equalTo(previousTextField.snp.bottom).offset(offset)
+            make.leading.equalTo(previousTextField.snp.leading)
+        }
+    }
+    
+    private func observeEmailField() {
+        Publishers.CombineLatest3(loginViewModel.$email, loginViewModel.isEmailEmptyPublisher, loginViewModel.isEmailValidPublisher)
+            .sink { [weak self] (_, isEmpty, isValid) in
+                guard let self = self else { return }
+                if isEmpty {
+                    self.updateConstraintsForEmptyTextField(self.passwordTextField, relativeTo: self.emailTextField, isEmpty: isEmpty)
+                    self.emptyEmailFieldWarning.isHidden = !isEmpty
+                    self.emailWarninigLabel.isHidden = isEmpty
+                } else {
+                    self.updateConstraintsForValidTextField(self.passwordTextField, relativeTo: self.emailTextField, isValid: isValid)
+                    self.emptyEmailFieldWarning.isHidden = !isEmpty
+                    self.emailWarninigLabel.isHidden = isValid
+                }
+            }
+            .store(in: &cancellables)
+    }
+    
+    private func observePasswordField() {
+        Publishers.CombineLatest3(loginViewModel.$password, loginViewModel.isPasswordEmptyPublisher, loginViewModel.isPasswordValidPublisher)
+            .sink { [weak self] (_, isEmpty, isValid) in
+                guard let self = self else { return }
```

Comment:
Теперь, помнится, можно даже так:
```suggestion
                guard let self else { return }
```